File: src/main/scala/simann/EightQueens.scala
package simann

import scalaz.effect.{IO, SafeApp}
import scalaz.effect.IO.putStrLn
import scalaz._, Scalaz._

case class Board(indicesAtRow: List[Int]) extends AnyVal {
  def size = indicesAtRow.size
  def swap(x: Int, y: Int) = Board(indicesAtRow.updated(x, indicesAtRow(y)).updated(y, indicesAtRow(x)))
  def hasQueen(row: Int, col: Int) = indicesAtRow(row) === col
}

object Board {
  def apply(qs: Int*): Board = {
    val qsList = qs.toList
    require(qsList.distinct.size === qsList.size)
    Board(qsList)
  }
  def clean(size: Int) = Board((0 until size).toList)

  implicit val boardHasStringRep = Show.shows((b: Board) => {
    val blankRow = List.fill(b.size)('.')
    val rows = b.indicesAtRow.toList.map(blankRow.updated(_, 'Q'))
    rows.map(_.mkString).mkString("\n")
  })

  implicit val boardIsAnnealable = new Annealable[Board] {
    def heat(board: Board) = for {
      x <- Random.nextInt(board.size - 1)
      y <- Random.nextInt(board.size - 1)
      heated <- (x === y) ? heat(board) | board.swap(x, y).pure[({type M[B] = State[Random, B]})#M]
    } yield heated

    def energy(b: Board) = {
      def countConflictsToWest(row: Int, col: Int, dy: Int): Int = {
        val diagonalPositions = unfold((row, col)) {
          case (r, c) => (r >= 0 && c >= 0 && r < b.size) ?? ((r, c), (r + dy, c - 1)).some
        }.toList
        diagonalPositions.isEmpty ? 0 | diagonalPositions.tail.count {
          case (r, c) => b.hasQueen(r, c)
        }
      }
      b.indicesAtRow.zipWithIndex.map {
        case (c, r) => countConflictsToWest(r, c, -1) + countConflictsToWest(r, c, 1)
      }.sum
    }
  }
}

object EightQueens extends SafeApp {
  override def runl(args: List[String]): IO[Unit] = for {
    random <- RandomIO.random
    boardSize = args.lift(0).map(_.toInt).getOrElse(8)
    result = produceSolution(boardSize).eval(random)
    _ <- putStrLn(result.map(_.shows).getOrElse("No solution found"))
  } yield ()

  def produceSolution(boardSize: Int): State[Random, Option[Board]] = for {
    board <- initialBoard(boardSize)
    temperatures = Annealing.temperatures(Temperature(50.0), Temperature(0.5), 0.99)
    solution <- Annealing.anneal(board, temperatures, 1000)
  } yield {
    solution
  }

  private[simann] def initialBoard(size: Int): State[Random, Board] = Looping.loopS(Board.clean(size))(size, Board.boardIsAnnealable.heat _)
}

File: src/main/hs/EightQueens.hs
{-# LANGUAGE ScopedTypeVariables #-}

module EightQueens where

import Annealing
import Control.Monad (join, liftM)
import Control.Monad.Random (Rand, RandomGen, getRandomR)
import Data.List (unfoldr)

newtype Board = Board [Int]

produceSolution :: (RandomGen r) => Int -> Rand r (Maybe Board)
produceSolution size = do
    board <- initialBoard size
    let temperatureValues = temperatures (Temperature 100.0) (Temperature 0.1) 0.99
    solution <- anneal board temperatureValues 1000
    return solution

instance Annealable Board where
    heat board = let s = size board - 1 in do
        x <- getRandomR (0, s)
        y <- getRandomR (0, s)
        let board' :: Board = swap board x y
        if (x == y) then heat board else return board'

    energy b@(Board bs) = fromIntegral $ sum $ map (\(c, r) -> (countConflictsToWest b r c (-1)) + (countConflictsToWest b r c 1)) (bs `zip` [0..])
        where countConflictsToWest :: Board -> Int -> Int -> Int -> Int
              countConflictsToWest b@(Board bs) row col dy = if (null diagonalPositions) then 0 else count (\(r, c) -> hasQueen b r c) (tail diagonalPositions)
                  where diagonalPositions :: [(Int, Int)] = unfoldr (\(r, c) -> if (r >= 0 && c >= 0 && r < size b) then Just $ ((r, c), (r + dy, c - 1)) else Nothing) (row, col)

instance Show Board where
    show b@(Board ps) = 
        let blankRow = replicate (size b) '.'
            rows = map (\i -> adjust (const 'Q') i blankRow) ps
        in join $ map (++ ['\n']) rows

size :: Board -> Int
size (Board ps) = length ps

swap :: Board -> Int -> Int -> Board
swap (Board bs) x y = 
    let bx = bs !! x
        by = bs !! y
        b1 = adjust (const bx) y bs
    in Board $ adjust (const by) x b1

hasQueen (Board ps) r c = (ps !! r) == c

clean size = Board [0..(size-1)]

initialBoard :: (RandomGen g) => Int -> Rand g Board
initialBoard size = loopM (clean size) size heat

adjust :: (a -> a) -> Int -> [a] -> [a]
adjust f i as = (take i as) ++ ((f (as !! i) : drop (i+1) as))

count p xs = length $ filter p xs